R/carehomes_ifr_t.R

```diff
@@ -76,89 +97,120 @@ carehomes_ifr_t <- function(step, S, I_weighted, p, type = NULL) {
   IFR_by_group_and_vacc_class <-
     carehomes_IFR_t_by_group_and_vacc_class(step, p)
   beta <- sircovid_parameters_expand_step(step, p$beta_step)
-  n_vacc_classes <- nlayer(p$rel_susceptibility)
-
 
   calculate_expected_infections <- function(t, no_vacc) {
     m <- p$m
     ages <- seq_len(p$n_age_groups)
-    ch <- seq(to = p$n_groups, length.out = 2)
+    ch <- seq(to = n_groups, length.out = 2)
     m[ages, ] <- beta[t] * m[ages, ]
     m[ch, ages] <- beta[t] * m[ch, ages]
 
-    m_extended <- matrix(t(matrix(m, p$n_groups, p$n_groups * n_vacc_classes)),
-                         p$n_groups * n_vacc_classes,
-                         p$n_groups * n_vacc_classes,
+    m_extended <- matrix(t(matrix(m, n_groups, n_groups * n_vacc_classes)),
+                         n_groups * n_vacc_classes,
+                         n_groups * n_vacc_classes,
                          byrow = TRUE)
 
     ## probability of infection in next time step by group and vaccine class
+    rel_sus <- p$rel_susceptibility
+    rel_inf <- p$rel_infectivity
     if (no_vacc) {
-      ## exclude vaccine effects
-      prob_infection <- (1 - exp(-p$dt * (m_extended %*% I_weighted[, t])))
-    } else {
-      ## include vaccine effects
-      if (length(dim(p$rel_susceptibility)) == 3) {
-        ## as in other places only look at strain 1
-        rel_sus <- p$rel_susceptibility[, 1, , drop = FALSE]
-        rel_inf <- p$rel_infectivity[, 1, , drop = FALSE]
-      }
-      prob_infection <- (1 - exp(-p$dt * c(rel_sus) *
-                     (m_extended %*% (I_weighted[, t] * c(rel_inf)))))
+      rel_sus[, , ] <- 1
+      rel_inf[, , ] <- 1
     }
 
     ## expected infections in next time step by group and vaccine class
-    expected_infections <-  S[, t] * prob_infection
+    if (n_strains == 1) {
+      ## force of infection by group and vaccine class (as a vector)
+      foi <- c(rel_sus) * (m_extended %*% (I_weighted[, t] * c(rel_inf)))
+      ## probability of infection by group and vaccine class
+      prob_infection <- (1 - exp(-p$dt * foi))
+      ## expected infections by group and vaccine class
+      expected_infections <- S[, t] * prob_infection
+    } else {
+      ## need to sum over metastrains
+      II <- array(I_weighted[, t], c(n_groups, n_strains, n_vacc_classes))
+      II[, 1, ] <- II[, 1, ] + II[, 4, ]
+      II[, 2, ] <- II[, 2, ] + II[, 3, ]
+      II <- II[, -c(3, 4), , drop = FALSE]
```

Comment:
could/should we add a check that n_strains is indeed 4 and n_real_strain indeed 2 here otherwise may run into issues with all these hard coded numbers?

Reply:
Added a check! n_strains = 4 is basically hardcoded into the multistrain version of the model too. But good idea, add this in so we know this needs changing if the multistrain part changes again